src/concurrent.rs

```diff
@@ -0,0 +1,170 @@
+// Copyright 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License.  You may obtain a copy
+// of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+use nix::unistd;
+use std::fs;
+use std::io;
+use std::os::unix::io as unix_io;
+use std::sync::{Arc, Mutex};
+
+/// A file with a single scoped owner but with multiple non-owner views.
+///
+/// A `ShareableFile` object owns the file passed to it at construction time and will close the
+/// underlying file handle when the object is dropped.
+///
+/// Concurrent views into this same file, obtained via the `clone_unowned` method, do not own the
+/// file handle.  Such views must accept the fact that the handle can be closed at any time, a
+/// condition that is simply exposed as if the handle reached EOF.  Concurrent views can safely be
+/// moved across threads.
```

Comment:
The reason is this: there will be two threads. One blocked on `fuse::mount` and another (the reconfiguration thread) blocked on a read from the input file.

Once `fuse::mount` terminates (because the user typed `umount foo` from the command line, for example), we want to terminate the reconfiguration thread so that we can collect its status and then exit. But to terminate the reconfiguration thread, we need to unblock the read. Closing the file descriptor does this.

You can poke through https://github.com/jmmv/sandboxfs/tree/reconfigs for the not-yet-ready-to-review code that uses this feature to get a better picture. In particular, see `src/lib.rs` and `src/config.rs`.

Reply:
I think I understand. It is a weird use case but this doesn't seem reasonable.

Are you sure that you want to return EOF in this case though? Why not return an error with a somewhat helpful value? It could be confusing if people "successfully" read a partial file.

Reply:
I think EOF makes sense in this case. That's how it'll be treated anyway by the caller in all cases. (Once the user has unmounted the file system, anything that comes through the reconfiguration channel doesn't matter at all. I'd just as well "leak" the thread, which is what the Go version does I think, but properly controlling termination seems nicer.)

Note that the way the reconfiguration protocol works right now is a huge hack that was put in place a while ago. For now I'm just mimicking the Go version of the code so that I can drop in the Rust replacement and evaluate performance -- but I want to redo this whole thing at a later stage and this code will hopefully just go away.

Reply:
Ok. I guess I was just thinking that in the case someone is reading the filesystem after unmounting it might be easier to debug if they get some sort of error instead of an EOF. But since you know the domain you know best.